### PR TITLE
Allow for homebrew/core to be untapped

### PR DESCRIPTION
**Pull request by Rylan12**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

This PR allows homebrew/core to be untapped if the `HOMEBREW_UNTAP_HOMEBREW_CORE` environment variable is set. We may want to consider merging this variable with `HOMEBREW_BOTTLE_JSON`. For now, I separated them because they are technically for different things (and I wasn't sure about a name). Happy to do whatever we think is best here.

Now, `Formulary` will throw a `CoreTapFormulaUnavailableError` when it can't find a formula under the conditions that homebrew/core is untapped and the environment variable is set.

`CoreTapFormulaUnavailableError` is a subclass of `TapFormulaUnavailableError` which is a subclass of `FormulaUnavailableError`. Therefore, anytime `CoreTapFormulaUnavailableError` is thrown it will still be caught by any `rescue FormulaUnavailableError` blocks. However, it will now also be caught by `resuce TapFormulaUnavailableError` blocks which means that more appropriate handling can be done. For example, we ignore missing `conflicts_with` formulae when a `TapFormulaUnavailableError` is raise but not when `FormulaUnavailableError` is raised. We want `CoreTapFormulaUnavailableError` to have the same logic as `TapFormulaUnavailableError` in this case, because the conflicts that are missing should be treated just as if they conflict with a different tap.

The other more major change is that `Dependency::expand` will now automatically prune all deps that raise `CoreTapFormulaUnavailableError` when calling `#to_formula`. This does mean that things like `brew deps` will not show any build/test dependencies that aren't installed locally, and may have other similar consequences.

I think this is okay, though, because the people who will have homebrew/core untapped likely won't need to worry about these things. Having homebrew/core untapped will definitely be bad for Homebrew developers, but I think that most of the weird side-effects that may appear won't matter for the general user. For example, a user who just wants to install a formula quickly won't care if running `brew deps --include-test --include-build` on that formula doesn't list all of the test/build deps that aren't installed already.

---

Here's how I tested:

```console
$ export HOMEBREW_BOTTLE_JSON=1

$ export HOMEBREW_UNTAP_HOMEBREW_CORE=1

$ brew untap homebrew/core
...
Untapping homebrew/core...
Untapped 2 commands and 5654 formulae (6,150 files, 435.5MB).

$ brew json rnv
==> Downloading https://ghcr.io/v2/homebrew/core/rnv/blobs/sha256:c262efcf45b880c131f5e466d1b672ce5120dff6302b9b7504f6c1e1ee87cb22
Already downloaded: /Users/rylanpolster/Library/Caches/Homebrew/downloads/22cd8e6434267f3561d6b0508f57ce5a83ac620b10b90b60f900abe2e4641b76--rnv--1.7.11.big_sur.bottle.tar.gz
==> Downloading https://ghcr.io/v2/homebrew/core/expat/blobs/sha256:18ebdde1356dd48de8c575eb2479dbb6556f7bef7a2b48a64c0f018dd79af1dc
Already downloaded: /Users/rylanpolster/Library/Caches/Homebrew/downloads/75221c8c9bb68848b82f46840d432a4b72bcf8318122376c3dce0b4ab1a081ec--expat--2.4.1.big_sur.bottle.tar.gz
==> Installing dependencies for rnv: expat
==> Installing rnv dependency: expat
==> Pouring expat--2.4.1.big_sur.bottle.tar.gz
🍺  /usr/local/Cellar/expat/2.4.1: 21 files, 568.7KB
==> Installing rnv
==> Pouring rnv--1.7.11.big_sur.bottle.tar.gz
🍺  /usr/local/Cellar/rnv/1.7.11: 10 files, 661.5KB

$ brew uninstall rnv expat
Uninstalling /usr/local/Cellar/rnv/1.7.11... (10 files, 661.5KB)
Uninstalling /usr/local/Cellar/expat/2.4.1... (21 files, 568.7KB)
```

Once the formula is installed, `Formulary::factory` will find it from the cellar without any issue, so things like `deps`, `uses --installed`, `info`, `desc`, `home`, `link`, `unlink`, `list`, `pin`, and `unpin` all work as expected (I tested all of those except the actual install logic with `pin` and `unpin`).

I haven't even bothered testing `brew upgrade` and `brew reinstall`, yet. Those will need some additional thought which will be one of the next steps.

Lastly, I did test installing a cask which seems to work fine. I haven't played with casks that depend on formulae at all yet. That will be something for much later.

### CODE REVIEW

**Comment by carlocab** [comment]
Minor suggestion; might be unnecessary depending on how long you plan on keeping the sub-command around: can we rename `json` to `json-install`?

**Comment by Rylan12** [comment]
> can we rename `json` to `json-install`?

Fine with me (that doesn't directly affect anything here).

Once this is merged I'm going to spend some time thinking about the migration plan. I know @MikeMcQuaid has some thoughts (and I think a potential plan in mind) but I don't think we're on the same page yet about that. I'll probably open an issue in homebrew/json to outline what needs to be done. That will include things like migrating the `json` (or `json-install`) command (or not migrating, depending), figuring out env variables and when they become documented/default (and for whom), handling `brew update`, `brew upgrade`, and `brew reinstall`, things like that.

**Comment by MikeMcQuaid** [comment]
> Minor suggestion; might be unnecessary depending on how long you plan on keeping the sub-command around: can we rename `json` to `json-install`?

Let's not. It's a temporary hack to speed up development.

> Once this is merged I'm going to spend some time thinking about the migration plan. I know @MikeMcQuaid has some thoughts (and I think a potential plan in mind) but I don't think we're on the same page yet about that. I'll probably open an issue in homebrew/json to outline what needs to be done. That will include things like migrating the `json` (or `json-install`) command (or not migrating, depending), figuring out env variables and when they become documented/default (and for whom), handling `brew update`, `brew upgrade`, and `brew reinstall`, things like that.

Sounds good! In case it helps my broad thinking is:
- get everything needed for a v1 (or v0.1) working in the `brew json` command that can take either JSON URLs or formulae names
- add logic guarded on undocumented `ENV["HOMEBREW_SOMETHING"].present?` inside Homebrew/brew (and have `brew json` set this) that allows Homebrew to be used for installation without Homebrew/core being tapped. 
- get maintainers testing Homebrew by setting `HOMEBREW_SOMETHING` and playing around with `brew json` and an untapped homebrew/core
- enable `HOMEBREW_SOMETHING` behaviour for everyone with `HOMEBREW_DEVELOPER` set and document it
- enabled `HOMEBREW_SOMETHING` functionality for everyone (but we auto-tap homebrew/core when you run a `dev-cmd` or have `HOMEBREW_DEVELOPER` set)

File: Library/Homebrew/cmd/untap.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/untap.rb:28]:
```suggestion
      odie "Untapping #{tap} is not allowed" if tap.core_tap? && ENV["HOMEBREW_UNTAP_HOMEBREW_CORE"].blank?
```

File: Library/Homebrew/dependency.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dependency.rb:142]: What's this doing? Wasn't obvious to me.

File: Library/Homebrew/formulary.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb:364]:
```suggestion
      if !CoreTap.instance.installed? && ENV["HOMEBREW_UNTAP_HOMEBREW_CORE"].present?
```

File: Library/Homebrew/tap.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/tap.rb:737]:
```suggestion
    return if ENV["HOMEBREW_UNTAP_HOMEBREW_CORE"].present?
```

File: Library/Homebrew/tap.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/tap.rb:754]:
```suggestion
    raise "Tap#uninstall is not available for CoreTap" if ENV["HOMEBREW_UNTAP_HOMEBREW_CORE"].blank?
```

File: Library/Homebrew/cmd/update-report.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update-report.rb:210]:
```suggestion
    return if ENV["HOMEBREW_UNTAP_HOMEBREW_CORE"].blank?
```

**Review by MikeMcQuaid — DISMISSED** [review]

**Comment by MikeMcQuaid** [comment]
> This PR allows homebrew/core to be untapped if the `HOMEBREW_UNTAP_HOMEBREW_CORE` environment variable is set. We may want to consider merging this variable with `HOMEBREW_BOTTLE_JSON`. For now, I separated them because they are technically for different things (and I wasn't sure about a name). Happy to do whatever we think is best here.

Yeh, let's have a maybe `HOMEBREW_JSON_CORE` variable and use that for both?

> We want `CoreTapFormulaUnavailableError` to have the same logic as `TapFormulaUnavailableError` in this case, because the conflicts that are missing should be treated just as if they conflict with a different tap.

👍🏻 

> The other more major change is that `Dependency::expand` will now automatically prune all deps that raise `CoreTapFormulaUnavailableError` when calling `#to_formula`. This does mean that things like `brew deps` will not show any build/test dependencies that aren't installed locally, and may have other similar consequences.
> 
> I think this is okay, though, because the people who will have homebrew/core untapped likely won't need to worry about these things.

This seems undesirable behaviour to me. In this case I'd think we should output dependencies entirely based on cached JSON and/or formulae files or just error out if we're unable to do so correctly. The `brew deps` `runtime_dependencies` paths might be most interesting because they allow doing the output based purely on a single `Tab`.

File: Library/Homebrew/dependency.rb
Comment (Rylan12) [reply Library/Homebrew/dependency.rb:142]: I'll add a comment. Essentially, ignore dependencies (and their recusrsive dependencies) that don't exist because homebrew/core is untapped.

For example, if this is the dependency structure of `foo` (where `bar` is a build only dependency):

```
$ brew deps --tree --include-build foo
foo
├── bar
│   └── baz
└── formula
```

Normally (with homebrew/coe tapped), here's what `Dependency.expand Formula["foo"]` returns:

```irb
[1] brew(main)> Dependency.expand Formula["foo"]
=> [#<Dependency: "baz" [:build]>,
 #<Dependency: "bar" [:build]>,
 #<Dependency: "formula" []>]
```

Now, with homebrew/core untapped (and `bar` and `baz` not installed):

```irb
[1] brew(main)> Dependency.expand Formula["foo"]
=> [#<Dependency: "formula" []>]
```

Functionally, this is okay because build dependencies aren't needing when installing from a bottle. Also, if `#to_formula` is ever called on one of the missing deps, it will still raise a `CoreTapFormulaUnavailableError`. These are only missing from `Dependency::expand`.

**Comment by Rylan12** [comment]
> > The other more major change is that `Dependency::expand` will now automatically prune all deps that raise `CoreTapFormulaUnavailableError` when calling `#to_formula`. This does mean that things like `brew deps` will not show any build/test dependencies that aren't installed locally, and may have other similar consequences.
> > I think this is okay, though, because the people who will have homebrew/core untapped likely won't need to worry about these things.
> 
> This seems undesirable behaviour to me. In this case I'd think we should output dependencies entirely based on cached JSON and/or formulae files or just error out if we're unable to do so correctly. The `brew deps` `runtime_dependencies` paths might be most interesting because they allow doing the output based purely on a single `Tab`.

You do run into an odd problem if you have casks installed that have any formula dependencies that, in turn, have build/test dependencies that you don't have installed.

This is because when uninstalling, brew checks to make sure that no installed casks depend on the formulae being uninstalled. For formulae, this is done using `Dependenct#missing_dependencies`. For casks, though, the runtime dependencies of the cask are all found and checked. So, even if its not going to matter for the uninstall checks, an error is still raised because `Dependency::expand` can't find some of the dependencies.

I've updated so that this pruning functionality is only done within the `InstalledDependents::find_some_installed_dependents` method, limiting it to this particular case.

**Review by MikeMcQuaid — DISMISSED** [review]
Looks good to me (once `brew style --fix` is run)! Nice work @Rylan12!

**Pull request merged**
